File: packages/sdkwork-utils-java/src/main/java/com/sdkwork/utils/BytesUtils.java
package com.sdkwork.utils;

import java.util.Locale;

public final class BytesUtils {
    private static final String[] UNITS = {"B", "KB", "MB", "GB", "TB", "PB"};

    private BytesUtils() {
    }

    public static String formatBytes(long bytes, int decimals) {
        long normalized = Math.max(0L, bytes);
        if (normalized < 1024L) {
            return normalized + " B";
        }

        double size = normalized;
        int unitIndex = 0;
        while (size >= 1024.0 && unitIndex < UNITS.length - 1) {
            size /= 1024.0;
            unitIndex++;
        }

        return String.format(Locale.ROOT, "%." + decimals + "f %s", size, UNITS[unitIndex]);
    }
}

File: packages/sdkwork-utils-kotlin/src/main/kotlin/com/sdkwork/utils/BytesUtils.kt
package com.sdkwork.utils

object BytesUtils {
    private val UNITS = arrayOf("B", "KB", "MB", "GB", "TB", "PB")

    fun formatBytes(bytes: Long, decimals: Int): String {
        val normalized = maxOf(0L, bytes)
        if (normalized < 1024L) {
            return "$normalized B"
        }

        var size = normalized.toDouble()
        var unitIndex = 0
        while (size >= 1024.0 && unitIndex < UNITS.lastIndex) {
            size /= 1024.0
            unitIndex++
        }

        return "%.${decimals}f ${UNITS[unitIndex]}".format(size)
    }
}
